src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,483 @@
+ *
+ *}
+ *
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public interface HKDFParameterSpec extends AlgorithmParameterSpec {
+
+    /**
+     * This {@code Builder} builds {@code Extract} and {@code ExtractThenExpand}
+     * objects.
+     * <p>
+     * The {@code Builder} is initialized via the {@code ofExtract} method of
+     * {@code HKDFParameterSpec}. As stated in the class description,
+     * {@code addIKM} and/or {@code addSalt} may be called as needed. Finally,
+     * the object is "built" by calling either {@code extractOnly} or
+     * {@code thenExpand} for {@code Extract} and {@code ExtractThenExpand}
+     * use-cases respectively. Note that the {@code Builder} is not
+     * thread-safe.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Builder {
+
+        List<SecretKey> ikms = new ArrayList<>();
+        List<SecretKey> salts = new ArrayList<>();
+
+        private Builder() {}
+
+        /**
+         * Creates a {@code Builder} for an {@code Extract}.
+         *
+         * @return a {@code Builder} to mutate
+         */
+        private Builder createBuilder() {
+            return this;
+        }
+
+        /**
+         * Builds an {@code Extract} from the current state of the
+         * {@code Builder}.
+         *
+         * @return an immutable {@code Extract}
+         */
+        public Extract extractOnly() {
+            return new Extract(ikms, salts);
+        }
+
+        /**
+         * Builds an {@code ExtractThenExpand}.
+         *
+         * @param info
+         *     the optional context and application specific information (may be
+         *     {@code null}); the byte array is copied to prevent subsequent
+         *     modification
+         * @param length
+         *     the length of the output key material
+         *
+         * @return an {@code ExtractThenExpand}
+         *
+         * @throws IllegalArgumentException
+         *     if {@code length} is not greater than 0
+         */
+        public ExtractThenExpand thenExpand(byte[] info, int length) {
+            return new ExtractThenExpand(
+                extractOnly(), info,
+                length);
+        }
+
+        /**
+         * Adds input key material to the builder.
+         * <p>
+         * {@code addIKM} may be called when the input key material value is to
+         * be assembled piece-meal or if part of the IKM is to be supplied by a
+         * hardware crypto device. This method appends to the existing list of
+         * values or creates a new list if there is none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the IKM
+         * but the actual value of the IKM is not yet available.
+         * <p>
+         * An implementation should concatenate the input key materials into a
+         * single value once all components are available.
```

Comment:
The above is a requirement for implementations and should be better to be moved to `ikms`. You can keep the line here with `would` instead of `should`.